SwiftCompilerSources/Sources/Optimizer/Analysis/LoopTree.swift

```diff
@@ -0,0 +1,209 @@
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+import OptimizerBridging
+
+/// Describes top level loops.
+struct LoopTree {
+  fileprivate let bridged: BridgedLoopTree
+  
+  let loops: TopLevelLoopArray
+  
+  init(bridged: BridgedLoopTree, context: FunctionPassContext) {
+    self.bridged = bridged
+    self.loops = TopLevelLoopArray(bridged)
+  }
+}
+
+/// Describes a loop with its children.
+struct Loop {
+  private let bridged: BridgedLoop
+  
+  let innerLoops: LoopArray
+  let loopBlocks: LoopBlocks
+  
+  var exitingAndLatchBlocks: some Sequence<BasicBlock> {
+    return header.predecessors.lazy
+      .filter { predecessor in
+        contains(block: predecessor) && !isLoopExiting(loopBlock: predecessor)
+      } + exitingBlocks
+  }
+  
+  /// Exit blocks of the loop.
+  ///
+  /// - Note: Some exit blocks will be duplicated if the loop has critical edges.
+  var exitBlocks: some Sequence<BasicBlock> {
+    return loopBlocks.lazy
+      .flatMap(\.successors)
+      .filter { !contains(block: $0) }
+  }
+  
+  var exitingBlocks: some Sequence<BasicBlock> {
+    return loopBlocks.lazy
+      .filter { isLoopExiting(loopBlock: $0) }
+  }
+  
+  init(bridged: BridgedLoop) {
+    self.bridged = bridged
+    self.innerLoops = LoopArray(bridged)
+    self.loopBlocks = LoopBlocks(bridged)
+  }
+  
+  var preheader: BasicBlock? {
+    bridged.getPreheader().block
+  }
+  
+  var header: BasicBlock {
+    bridged.getHeader().block
+  }
+  
+  /// Returns `true` if the loop has exactly one exit block.
+  var hasSingleExitBlock: Bool {
+    return exitBlocks.singleElement != nil
+  }
+  
+  var hasNoExitBlocks: Bool {
+    return exitBlocks.isEmpty
+  }
+  
+  private func isLoopExiting(loopBlock: BasicBlock) -> Bool {
+    return loopBlock.successors.contains { !contains(block: $0) }
+  }
+  
+  func getBlocksThatDominateAllExitingAndLatchBlocks(domTree: DominatorTree) -> some Sequence<BasicBlock> {
+    return getBlocksThatDominateAllExitingAndLatchBlocksHelper(bb: header, domTree: domTree)
+  }
+  
+  func contains(block: BasicBlock) -> Bool {
+    return bridged.contains(block.bridged)
+  }
+
+  private func getBlocksThatDominateAllExitingAndLatchBlocksHelper(
+    bb: BasicBlock,
+    domTree: DominatorTree
+  ) -> some Sequence<BasicBlock> {
+    guard exitingAndLatchBlocks.allSatisfy({ exitBlock in
+      return bb.dominates(exitBlock, domTree)
+    }) else {
+      return []
+    }
+    
+    return [bb] + domTree.getChildren(of: bb).lazy
+      .flatMap { child in
+        getBlocksThatDominateAllExitingAndLatchBlocksHelper(
+          bb: child,
+          domTree: domTree
+        )
+      }
+  }
+  
+  func splitCriticalEdges(_ context: FunctionPassContext) {
```

Comment:
A better name would be `splitCriticalExitingAndBackEdges` because it's not splitting critical edges within the loop